### DISCUSSION TITLE
Inhibit screen saver while connected?

### DISCUSSION THREAD

**Discussion by whot** [input-leap/input-leap#1707] [Development of Input Leap]
Broken out from #1699 

One of the problems with the libei backend (read: wayland) is that the screensaver can switch on and will kick us out of the RemoteDesktop session. Reconnecting is possible (see #1705) but that requires disabling the screensaver manually somehow.

That the screensaver switches on while we are connected is arguably a problem that needs to be fixed separately. We could use the [Inhibit portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Inhibit) to prevent this, see #1706. That's a security issue though since the screen will never lock while input-leap is *running*.

The alternative is enable RemoteDesktop to work while idle/screen is locked, but I don't know how and whether that is possible atm.

cc @prahal, @jadahl

**Comment by jadahl** [comment]
Tricky... the closing of the session is an intentional thing to make it impossible to accidentally share your screen when you're accessing your session remotely. Here, "remotely" is within a meter so it doesn't have the same issue, but that isn't possible to know in the portal.

**Reply by p12tic** [reply]
That's unfortunate. A data point from my personal perspective as a user is that it's quite annoying to unlock the screen using other means (in my case - physical keyboard and mouse which is also nearby). This is just one data point though.

**Reply by jadahl** [reply]
I can imagine that, just need to come up with a solution here that to some degree avoids people accessing via e.g. RDP actually remotely doesn't unknowingly unlock their physical screen in a place they are not at.

@whot had an idea with some "pass phrase" shown on screen that one has to enter e.g. "Please enter 1 2 3 4" in addition to the actual unlock screen, but it has some flaws too, but maybe we can figure out a way to make that safe still.

Perhaps something like these loosely thought of rules:

 * With a input-only remote desktop session, with a capability flag, allow keeping the session alive when the screen is locked
 * If something wants to unlock, show the "Please enter 1 2 3 4" on the physical monitor, but not via any screen recording
 * Only allow entering 1 2 3 4 from the remote desktop session InputLeap uses
 * If one touches a physical key, or e.g. clicks with a physical mouse, disconnect the remote session
 * Only allow creating remote sessions if a user session is already unlocked
 * Use session-restore token in InputLeap to transparently re-establish remote desktop sessions once unlocked, if it was thrown out

It does mean input leaps needs to be able to listen on some "screen is locked" signal; can't remember if this is a thing yet in portal land.

**Reply by whot** [reply]
> If one touches a physical key, or e.g. clicks with a physical mouse, disconnect the remote session

I think this bit is unnecessary, at least my use case here (admittedly mostly testing rather than *using* it) changes between remote and local devices frequently so disconnecting would be a tad annoying.

> It does mean input leaps needs to be able to listen on some "screen is locked" signal; can't remember if this is a thing yet in portal land.

There's a `StateChanged` signal in the `Inhibit` portal, maybe that does what we want.

**Reply by jadahl** [reply]
> I think this bit is unnecessary, at least my use case here (admittedly mostly testing rather than using it) changes between remote and local devices frequently so disconnecting would be a tad annoying.

My thought was that if you do touch the physical keyboard when the lock screen is active, your intention was to unlock it. And to avoid frustration after unlocking, have InputLeap re-establish the session transparently with a restore token.

**Reply by whot** [reply]
> your intention was to unlock it. 

possibly niche counterexample: volume keys

**Reply by jadahl** [reply]
Maybe niche enough to ignore...

**Reply by prahal** [reply]
SInce end of October, the input leap client unblanks when I move the cursor to it from the input leap server
https://github.com/input-leap/input-leap/issues/1699#issuecomment-1769643174
I currently have xdg-desktop-portal 1.18.2.

As I told into #1706 I am not fond of inhibiting the session (ie no screen blank anymore).

Sorry, I was not able to report earlier as I switched back to Xorg in the meantime.

But from your comments above it was a feature to kill RemoteDestop on the remote client, ie " the closing of the session is an intentional thing to make it impossible to accidentally share your screen when you're accessing your session remotely". I admit I do not understand the meaning of this sentence.
To me it  means to kill the RemoteDesktop Session when the screen blanks, to prevent us from sharing our screen (locally?) when we are accessing our session remotely.

Either way, the input leap client screen unblanks and I can input into it when I move the input leap server has its mouse pointer moving to it.

**Reply by whot** [reply]
> Either way, the input leap client screen unblanks and I can input into it when I move the input leap server has its mouse pointer moving to it.

I don't know how you're reproducing this, is this an upstream mutter change? With Fedora 39 the screen is locked and we lose the session.
